src/clinical/service-worker-thread/tasks.ts

```diff
@@ -149,14 +139,33 @@ const mapEntityDocuments = (
 	const { page, pageSize = results.length, sort } = paginationQuery;
 	const relevantSchemaWithFields = schemas.find((s: any) => s.name === entityName);
 	const entityInQuery = isEntityInQuery(entityName, entityTypes);
-	const entityErrors = errors.filter((errorRecord) => errorRecord.entityName === entityName);
 
 	if (!relevantSchemaWithFields || !entityInQuery || isEmpty(results)) {
 		return undefined;
 	}
 
 	const totalDocs = entityName === ClinicalEntitySchemaNames.DONOR ? donorCount : results.length;
-	let records = results.sort(sortDocs(sort, completionStats, sortType, entityErrors));
+
+	const sortKey = sort[0] === '-' ? sort.split('-')[1] : sort;
+	const key = sortKey === 'donorId' ? DONOR_ID_FIELD : sortKey;
+
+	const entityErrors =
+		sortType === ClinicalDataSortTypes.invalidEntity
+			? errors.filter((errorRecord) => errorRecord.entityName === entityName)
+			: [];
+	const donorIdsWithErrors =
+		sortType === ClinicalDataSortTypes.invalidEntity
+			? entityErrors.reduce<Set<number>>((acc, error) => acc.add(error.donorId), new Set())
+			: new Set([]);
+
+	const sortRecords =
+		sortType === ClinicalDataSortTypes.defaultDonor
+			? sortDocs(sort, sortDonorRecordsByCompletion, completionStats)
+			: sortType === ClinicalDataSortTypes.invalidEntity
+			? sortDocs(sort, sortInvalidRecords, donorIdsWithErrors)
+			: sortDocs(sort, sortRecordsByColumn, key);
```

Comment:
This is so hard to follow with all the ternaries... could we instead just do a switch on the different sort types and have the logic for each contained in its own block?

Reply:
Switch/case solution w/ TS generics to solve some of the issues noted above
https://github.com/icgc-argo/argo-clinical/pull/1184/commits/532c648c059826a5e35ab25ef8df8dcf40841307